common/src/tor.rs

```diff
@@ -0,0 +1,204 @@
+        instance.temp_dir = Some(temp);
+        instance.data_dir = dir;
+
+        // log destination
+        let temp = NamedTempFile::new()?.into_temp_path();
+        let file = temp.to_string_lossy().to_string();
+        instance.temp_file = Some(temp);
+        instance.log_destination = file;
+
+        Ok(instance)
+    }
+
+    /// Override a given Tor comms transport with the control address and auth from this instance
+    pub fn updated_comms_transport(&self, transport: CommsTransport) -> Result<CommsTransport, ExitCodes> {
+        if let CommsTransport::TorHiddenService {
+            socks_address_override,
+            forward_address,
+            auth,
+            onion_port,
+            tor_proxy_bypass_addresses,
+            tor_proxy_bypass_for_outbound_tcp,
+            ..
+        } = transport
+        {
+            let control_server_address = format!("/ip4/127.0.0.1/tcp/{}", self.control_port).parse::<Multiaddr>()?;
+            let auth = if let Some(ref passphrase) = self.passphrase {
+                TorControlAuthentication::Password(passphrase.to_owned())
+            } else {
+                auth
+            };
+            let transport = CommsTransport::TorHiddenService {
+                control_server_address,
+                socks_address_override,
+                forward_address,
+                auth,
+                onion_port,
+                tor_proxy_bypass_addresses,
+                tor_proxy_bypass_for_outbound_tcp,
+            };
+            Ok(transport)
+        } else {
+            let e = format!("Expected a TorHiddenService comms transport, received: {:?}", transport);
+            Err(ExitCodes::ConfigError(e))
+        }
+    }
+
+    /// Run the Tor instance until the shutdown signal is received
+    pub async fn run(self, mut shutdown_signal: ShutdownSignal) -> Result<(), ExitCodes> {
+        info!(target: LOG_TARGET, "Starting Tor");
+
+        let Tor {
+            data_dir,
+            socks_port,
+            control_port,
+            log_level,
+            log_destination,
+            passphrase,
+            ..
+        } = self;
+
+        let mut tor = libtor::Tor::new();
+
+        tor.flag(TorFlag::DataDirectory(data_dir.clone()))
+            .flag(TorFlag::SocksPort(socks_port))
+            .flag(TorFlag::ControlPort(control_port))
+            .flag(TorFlag::Hush())
+            .flag(TorFlag::LogTo(log_level, LogDestination::File(log_destination)));
+
+        if let Some(passphrase) = passphrase {
+            let hashed_pass = hashed_control_password(passphrase)?;
+            tor.flag(TorFlag::HashedControlPassword(hashed_pass));
+        }
+
+        tor.start_background();
+
+        loop {
+            tokio::select! {
+                _ = shutdown_signal.wait() => {
+                    info!(target: LOG_TARGET, "Shutdown signal received, dropping Tor");
+                    break;
+                }
+            }
+        }
```

Comment:
ShutdownSignal is just a future that resolves when triggered

```suggestion
                shutdown_signal.wait().await;
                    info!(target: LOG_TARGET, "Shutdown signal received, dropping Tor");    
```

Reply:
Thanks I knew I was missing something lol